src/shared/Atlassian.Bitbucket.Tests/DataCenter/BitbucketOAuth2ClientTest.cs

```diff
@@ -0,0 +1,149 @@
+
+            var client = GetBitbucketOAuth2Client();
+
+            MockGetAuthenticationCodeAsync(remoteUrl, rootCallbackUri, finalCallbackUri, clientId, client.Scopes);
+
+            MockCodeGenerator();
+
+            var result = await client.GetAuthorizationCodeAsync(browser.Object, ct);
+
+            VerifyAuthorizationCodeResult(result, rootCallbackUri);
+        }
+
+        [Fact]
+        public async Task BitbucketOAuth2Client_GetAuthorizationCodeAsync_ReturnsCode_WhileRespectingRedirectUriOverride()
+        {
+            var rootCallbackUrl = MockRootCallbackUriOverride("http://localhost:12345/");
+            var remoteUrl = MockRemoteUri("http://example.com");
+            var clientId = MockClientIdOverride("dc-client-id");
+            MockClientSecretOverride("dc-client-seccret");
+
+            Uri finalCallbackUri = MockFinalCallbackUri(new Uri(rootCallbackUrl));
+
+            var client = GetBitbucketOAuth2Client();
+
+            MockGetAuthenticationCodeAsync(remoteUrl, new Uri(rootCallbackUrl), finalCallbackUri, clientId, client.Scopes);
+
+            MockCodeGenerator();
+
+            var result = await client.GetAuthorizationCodeAsync(browser.Object, ct);
+
+            VerifyAuthorizationCodeResult(result, new Uri(rootCallbackUrl));
+        }
+
+        private void VerifyAuthorizationCodeResult(OAuth2AuthorizationCodeResult result, Uri redirectUri)
+        {
+            Assert.NotNull(result);
+            Assert.Equal(authorization_code, result.Code);
+            Assert.Equal(redirectUri, result.RedirectUri);
+            Assert.Equal(pkceCodeVerifier, result.CodeVerifier);
+        }
+
+        private Bitbucket.DataCenter.BitbucketOAuth2Client GetBitbucketOAuth2Client()
+        {
+            var client = new Bitbucket.DataCenter.BitbucketOAuth2Client(httpClient.Object, settings.Object, trace.Object);
+            client.CodeGenerator = codeGenerator.Object;
+            return client;
+        }
+
+        private void MockCodeGenerator()
+        {
+            codeGenerator.Setup(c => c.CreateNonce()).Returns(nonce);
+            codeGenerator.Setup(c => c.CreatePkceCodeVerifier()).Returns(pkceCodeVerifier);
+            codeGenerator.Setup(c => c.CreatePkceCodeChallenge(OAuth2PkceChallengeMethod.Sha256, pkceCodeVerifier)).Returns(pkceCodeChallenge);
+        }
+
+        private void MockGetAuthenticationCodeAsync(string url, Uri redirectUri, Uri finalCallbackUri, string overrideClientId, IEnumerable<string> scopes)
+        {
+            var authorizationUri = new UriBuilder(url + "/rest/oauth2/latest/authorize")
+            {
+                Query = "?response_type=code"
+             + "&client_id=" + (overrideClientId ?? "clientId")
+             + "&state=12345"
+             + "&code_challenge_method=" + OAuth2Constants.AuthorizationEndpoint.PkceChallengeMethodS256
+             + "&code_challenge=" + WebUtility.UrlEncode(pkceCodeChallenge).ToLower()
```

Comment:
I think it might just be a side-effect of the test? Without this the mock fails to match against the call to `GetAuthenticationCodeAsync` I'm not sure why yet

Reply:
Let me know if you think this is important to clarify.